src/main/java/org/springframework/data/couchbase/config/AbstractCouchbaseConfiguration.java

```diff
@@ -306,20 +311,30 @@ private ObjectMapper getCouchbaseObjectMapper() {
 	}
 
 	/**
-	 * Creates a {@link ObjectMapper} for the jsonSerializer of the ClusterEnvironment
+	 * Creates a {@link ObjectMapper} for the jsonSerializer of the ClusterEnvironment and spring-data-couchbase
+	 * jacksonTranslationService and also some converters (EnumToObject, StringToEnum, IntegerToEnum)
 	 *
 	 * @return ObjectMapper
 	 */
 	public ObjectMapper couchbaseObjectMapper() {
-		ObjectMapper om = new ObjectMapper(); // or use the one from the Java SDK (?) JacksonTransformers.MAPPER
-		om.configure(com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+		ObjectMapper om = getObjectMapper();
+		om.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
 		om.registerModule(new JsonValueModule());
 		if (getCryptoManager() != null) {
 			om.registerModule(new EncryptionModule(getCryptoManager()));
 		}
 		return om;
 	}
 
+	ObjectMapper objectMapper;
+
+	ObjectMapper getObjectMapper() {
```

Comment:
A few concerns:

* `couchbaseObjectMapper` and `getObjectMapper` do not appear to be thread-safe. Is that okay?

* Beacuse `getObjectMapper` caches the instance, `couchbaseObjectMapper` registers additional modules each time it is called.

* Again, because the instance is cached, the ObjectMapper returned by `getObjectMapper` might or might not have the configuration that was applied by couchbaseObjectMapper().

ObjectMapper is considered "thread-safe after configuration". It's not obvious from looking at the code that all calls to `couchbaseObjectMapper()` (which configures the mapper) happen before the mapper is used.


Suggestion:

* Remove the getObjectMapper() method, and call couchbaseObjectMapper() instead. Make couchbaseObjectMapper() responsible for caching the instance.

* Make the `objectMapper` field volatile.

* Inside `couchbaseObjectMapper()`, use double-checked locking to initialize the field if necessary. (This idiom is safe if the field being intialized is volatile.)
```
if (objectMapper == null) {
    synchronized {
        if (objectMapper == null) {
            objectMapper = new ObjectMapper();
            // configure it
        }
    }
}
```

Reply:
Something went off the rails here.   There was already a mapper object for caching.  Fix is coming.

The intention is to separate the caching from the initialization - as the user can override the initialization, and I don't want to burden them (or trust them) with the caching.  This would normally be an @Bean in the configuration, but it cannot be (there's a previous issue for that).  It is similar to cryptoManager which also cannot be an @Bean (because it can be null). 

 It should follow the same pattern as cryptoManager : getCryptoManager()  provides the cached cryptoManager, if necessary calling cryptoManager() which the user can override to create the cryptoManager.

So getObjectManager() provides the cached objectMapper, if necessary calling couchbaseObjectMapper() which the user may override for customization.

couchbaseObjectMapper() should be called only once.  I would make it private, but a user's override may want to call super.couchbaseObjectMapper(). (this is actually in one of the test cases).